packages/block-editor/src/components/global-styles/border-panel.js

```diff
@@ -187,38 +188,45 @@ export default function BorderPanel( {
 		// Ensure we have a visible border style when a border width or
 		// color is being selected.
 		const newBorderWithStyle = applyAllFallbackStyles( newBorder );
+		let updatedBorder = newBorderWithStyle;
 
-		// As we can't conditionally generate styles based on if other
-		// style properties have been set we need to force split border
-		// definitions for user set border styles. Border radius is derived
-		// from the same property i.e. `border.radius` if it is a string
-		// that is used. The longhand border radii styles are only generated
-		// if that property is an object.
-		//
-		// For borders (color, style, and width) those are all properties on
-		// the `border` style property. This means if the theme.json defined
-		// split borders and the user condenses them into a flat border or
-		// vice-versa we'd get both sets of styles which would conflict.
-		const updatedBorder = ! hasSplitBorders( newBorderWithStyle )
-			? {
-					top: newBorderWithStyle,
-					right: newBorderWithStyle,
-					bottom: newBorderWithStyle,
-					left: newBorderWithStyle,
-			  }
-			: {
-					color: null,
-					style: null,
-					width: null,
-					...newBorderWithStyle,
-			  };
+		if ( forceSplitBorders ) {
+			// As Global Styles can't conditionally generate styles based on if
+			// other style properties have been set, we need to force split
+			// border definitions for user set global border styles. Border
+			// radius is derived from the same property i.e. `border.radius` if
+			// it is a string that is used. The longhand border radii styles are
+			// only generated if that property is an object.
+			//
+			// For borders (color, style, and width) those are all properties on
+			// the `border` style property. This means if the theme.json defined
+			// split borders and the user condenses them into a flat border or
+			// vice-versa we'd get both sets of styles which would conflict.
+			updatedBorder = ! hasSplitBorders( newBorderWithStyle )
+				? {
+						top: newBorderWithStyle,
+						right: newBorderWithStyle,
+						bottom: newBorderWithStyle,
+						left: newBorderWithStyle,
+				  }
+				: {
+						color: null,
+						style: null,
+						width: null,
+						...newBorderWithStyle,
+				  };
+		}
 
-		[ 'top', 'right', 'bottom', 'left' ].forEach( ( side ) => {
-			updatedBorder[ side ] = {
-				...updatedBorder[ side ],
-				color: encodeColorValue( updatedBorder[ side ]?.color ),
-			};
-		} );
+		if ( hasSplitBorders( updatedBorder ) ) {
```

Comment:
So it's still possible to have split borders or not for the "block" case. In other words, the blocks are forced to understand the split borders, so why do we have an issue for the table block do you know?

Reply:
IIRC there was a decision made that setting differing borders per side on the Table block only made sense on the table as a whole not individual cells.

Reply:
Ok I confirmed my hunch. If I use split borders in the table block, the updates are not reflected on the UI, unless all the values are equal.

Reply:
Actually, they are reflected in the UI, it's just the 0 border that is wrong. So in other words, I think a better fix is to figure out why the 0 border is not reflected in the UI in the table block?

Reply:
> Actually, they are reflected in the UI, it's just the 0 border that is wrong. So in other words, I think a better fix is to figure out why the 0 border is not reflected in the UI in the table block?

The table is one issue. I recall there being some edge cases with the split border definitions and global styles when some things were defined as flat borders or not. I'm out of time today but would like to look into that history tomorrow to make sure that we don't introduce another regression.

Reply:
yeah, but If I'm not wrong, the component always uses split borders that way global styles continue to work. And blocks are supposed to understand split borders already. (you can set them in the UI), so what I'm saying is that there's a bug in the Table block when the user uses "split borders" regardless of my refactoring PR.

Reply:
> the component always uses split borders

I'm not sure I'm following. I'll need to revisit this tomorrow to catch up. This conditional has an else clause handling the flat border's color slug to align with your changes in the panel refactor. When you edit a flat border on an individual block this panel gets a flat non-split border.

FWIW, previously, individual blocks could have a simple flat border that reduces the inline styles in the block markup. Now, after the panel refactor, split borders have been forced so that's always full longhand for each property. That's a step back right?

Reply:
I'm saying that on the block, you click the "split" icon and actually use split borders, and putting different values in each side forces the component to use the "long format". So the block needs to understand that format too regardless.

Here's the bug


https://user-images.githubusercontent.com/272444/230041506-5fda5971-97bd-4938-bdf9-2b3e4b1597ee.mov

As you can see, I can set split borders with random values like 10 and it works properly, but I can't set split borders with the 0 value. 0 only works if I use flat border.

Reply:
Thanks for explaining further 👍️
> i’m saying that on the block, you click the "split" icon and actually use split borders, and putting different values in each side forces the component to use the "long format".


Understood. I agree this PR isn’t a complete fix for the Table block. It appears as though if you define a subset of side definitions with zero width, they are being filtered out or not applied to the table. I’ll look into that one tomorrow as well.

As noted above, the table borders are just one example issue after the recent forcing of split borders.

There’s still the issue of the bloated CSS forcing separate longhand values creates. I’m not aware of the style engine having the ability to merge longhand styles to shorthand yet. I imagine that optimisation has been considered and is on the cards.

> So the block needs to understand that format too regardless.


I agree that the Table block needs some level of understanding of the split definitions. I don’t think that’s true for all blocks though. Most individual blocks are oblivious to whether they have long or shorthand definitions. The border object is a property on their style object that is processed by the style engine and they’re given whatever classes and styles needed. 

So to me, it’s here when processing the border change that we should be keeping a flat border configuration when possible. Which then allows the output and saving of minimal inline styles. There’s still the original edge case I need to dig up to explain what forced the global styles to rely on splitting the definitions to the longhand value in the first place. 

If we can extend the style engine to be able to handle converting longhand rules to shorthand where possible, I'd be all for forcing split border definitions across the board.

Reply:
Let's be clear I don't mind keeping the "flat" format when possible but I don't like the flag we pass to make an exception for global styles. I think the "style" object from global styles and blocks should be as similar as possible. (We're obviously not there in a lot of things). 

At the same time, I don't think using long format consistently (which my PR did) created any bug in the table block, the bug already existed, it was just manifesting in less cases and the fix to that bug is not really reverting to short syntax. The fix is closer to this https://github.com/WordPress/gutenberg/pull/49603?notification_referrer_id=NT_kwDOAAQoPLE2MDcxNDI2NzAwOjI3MjQ0NA#discussion_r1158232779

Now if you feel strongly that we should land the current PR as a stop gap to be back to the previous state. I won't block it because it's minor but I still feel that it's not ideal.

Reply:
I've moved the logic forcing split borders to the Global Styles' border panel's `onChange` avoiding the introduction of any new prop (https://github.com/WordPress/gutenberg/pull/49603/commits/854ab4be06c3cc21ebf8a13c6f8fd348c2314914).

The PR description has been updated to focus only on the return to a reduced number of CSS styles being inlined. I'm working on a fix for the table block separately, so I think we can move forward and land this one.